### ISSUE TITLE
cargo test --doc not load same shared libraries as cargo test --lib

### ISSUE THREAD

**Issue by adamcrume** [rust-lang/cargo#8531]
**Problem**
`cargo test --doc` doesn't appear to be loading the same shared libraries as `cargo test --lib` in 1.45

**Steps**
1. Clone https://github.com/tensorflow/rust
2. Set Rust to 1.44
3. Note that both `cargo test --doc` and `cargo test --lib` pass
4. Set Rust to 1.45
5. Note that `cargo test --lib` passes but `cargo test --doc` fails with `error while loading shared libraries: libtensorflow.so.1: cannot open shared object file: No such file or directory`

**Notes**

Output of `cargo version`:
cargo 1.44.0 (05d080faa 2020-05-06) or cargo 1.45.0 (744bd1fbb 2020-06-15)

Running on Ubuntu 20.04 LTS.

**Comment by thoren-d** [comment]
For what it's worth, I'm seeing the same issue running doctests against my library that uses SDL2 (https://github.com/Rust-SDL2/rust-sdl2). `cargo +1.44.0 test --doc` works.

**Comment by raskr** [comment]
Same problem with macOS + rust1.48+, but not with rust1.42.
`cargo test --doc` seems to be broken.

**Comment by adamcrume** [comment]
Using `git bisect`, I narrowed down to 3fd28143de72d2d0e63186273d68aa890d8297e7 being the breaking commit.

**Comment by pipehappy1** [comment]
This is still broken on 1.58

**Comment by Trolldemorted** [comment]
This issue is known for almost 3 years, could you give us a timeline when a fix will come?

**Comment by justsmth** [comment]
Is there a workaround for this issue?

**Comment by Trolldemorted** [comment]
@adamcrume do you have any clue what might have caused this bug? Is @alexcrichton still around? Iirc he was stepping down from several Rust projects, but not cargo.

**Comment by adamcrume** [comment]
I don't know what specifically in 3fd28143de72d2d0e63186273d68aa890d8297e7 caused this.  I'm not familiar with the cargo codebase.

**Comment by weihanglo** [comment]
We appreciate Alex efforts on Cargo and Alex has already retired from the Cargo Team.[^1] 

There was a previous attempt <https://github.com/rust-lang/cargo/pull/10469> and concerns were listed <https://github.com/rust-lang/cargo/pull/10469#issuecomment-1079503079>. Feel free to pick it up and continue.

[^1]: https://blog.rust-lang.org/inside-rust/2022/03/31/cargo-team-changes.html

**Comment by Trolldemorted** [comment]
@weihanglo with all due respect, the concerns you linked do not point towards a, obvious, clear path forward how this issue should be remediated. I do understand you rely on others to help you fix bugs, but a community can do that with reasonable efficacy only if it is crystal clear **how** things should work. The concerns you listed are referenced and thought through, but they don't enable outsiders to grasp how you'd want things to be done.

This bug is a problem for the Rust ecosystem. I'd love to recommend Rust more strongly in industrial projects, but this is one of the reasons why I cannot. The cargo team has broken the test coverage for all crates that include native dylibs, and some crates have responded by disabling their doctests with `no_run`, other crates have stopped efforts to link against included native libraries with a guaranteed matching version, and instead rely on package managers installing a more or less matching version, thus causing undefined behavior. Cargo should not ignore bugs that incentivize maintainers to reduce test coverage or cause undefined behaviour for years.

@ehuss I hate to be nagging, but at which point in time will the cargo team fix this issue, if no volunteers are able to address it in a way you'd accept it? How do you select which bug fixes are scheduled for a new cargo release?

**Comment by tompscanlan** [comment]
Looks as though no one is looking at this.  I'm new to the code, but this seems approachable to me.  I'll take a wack if it's really open.

**Comment by justsmth** [comment]
EDIT: Sorry I hadn't yet read the full release announcement: https://blog.rust-lang.org/2025/08/07/Rust-1.89.0/#cross-compiled-doctests

----
Was there potentially a regression on this in Rust 1.89?  Our CI tests started failing around the time it released.

On some platforms, our "FIPS" build produces a shared library, and on those platforms the doctests (and only the doctests) are now failing:

https://github.com/aws/aws-lc-rs/actions/runs/16803801529/job/47591065792?pr=822
```
---- aws-lc-rs/src/aead.rs - aead (line 123) stdout ----
Test executable failed (exit status: 127).

stderr:
Error loading shared library libaws_lc_fips_0_13_7_crypto.so: No such file or directory (needed by /tmp/rustdoctestf6Vum3/rust_out)
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_AEAD_CTX_open_gather: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_AEAD_CTX_init: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_aead_aes_128_gcm: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_AEAD_CTX_seal_scatter: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_AEAD_CTX_seal: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_OPENSSL_malloc: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_AEAD_CTX_free: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_AEAD_CTX_zero: symbol not found
Error relocating /tmp/rustdoctestf6Vum3/rust_out: aws_lc_fips_0_13_7_EVP_AEAD_CTX_open: symbol not found
```

**Comment by mxgrey** [comment]
> on those platforms the doctests (and only the doctests) are now failing:

I'm encountering a somewhat similar issue. After attempting to upgrade from edition 2021 to 2024, I now get this error when I run `cargo test` from the virtual workspace:

```
/tmp/rustdoctest2f6aAq/rust_out: error while loading shared libraries: libstd-210854cf1daa4bec.so: cannot open shared object file: No such file or directory
all doctests ran in 3.45s; merged doctests compilation took 3.25s
```

It happens for the doctests of just one crate in my workspace. It happens whether I set `resolver = "2"` or `resolver = "3"` in the virtual workspace root manifest, but it goes away when I set the edition of the offending crate back to 2021.

Some additional peculiar details:
* If I go into the directory of the failing crate and run `cargo test` from inside there, the error doesn't happen. 
* If I run `cargo test -p <failing_crate>` then the error doesn't happen and the doctest passes
* Setting the offending crate back to edition 2021 allows `cargo test` to work from the virtual workspace root, without changing anything else. The resolver is still set to 3, and all other crates are set to edition 2024.
